app.go

```diff
@@ -621,3 +606,30 @@ func (a *Application) ReadOnlyState() appstate.State {
 		a.GetValidatorSet,
 	)
 }
+
+func (a *Application) InMemoryApp(blockNumber uint64, blockstore store.BlockStore) (middleware.InMemoryApp, error) {
+	startVersion := int64(blockNumber) - 1
+	if startVersion < 0 {
+		return nil, errors.Errorf("invalid block number %d", blockNumber)
+	}
+	for ; !a.Store.VersionExists(startVersion) || startVersion <= int64(0); startVersion-- {
+	}
+	if startVersion == 0 {
+		return nil, errors.Errorf("no saved version for height %d", blockNumber)
+	}
+	startStore, err := a.Store.RetrieveVersion(startVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	return middleware.NewInMemoryApp(
```

Comment:
This should really return a clone of `Application` with the snapshoted state, without all the irrelevant bits (e.g. `Application.QueryHandler`)